### ISSUE TITLE
Python AST deprecations

### ISSUE THREAD

**Issue by akaihola** [ikamensh/flynt#195]
In [my CI build](https://github.com/akaihola/darker/actions/runs/8318028688/job/22759492351?pr=563#step:8:50), I'm getting these deprecation warnings from Python 3.12:
```python
  lib/python3.12/site-packages/astor/op_util.py:92: DeprecationWarning:
    ast.Num is deprecated and will be removed in Python 3.14; use ast.Constant instead
    precedence_data = dict((getattr(ast, x, None), z) for x, y, z in op_data)
```
```python
  lib/python3.12/site-packages/flynt/utils/utils.py:74: DeprecationWarning:
    ast.Str is deprecated and will be removed in Python 3.14; use ast.Constant instead
    ) -> Union[ast.FormattedValue, ast.Str]:
```
```python
  lib/python3.12/site-packages/flynt/transform/format_call_transforms.py:22: DeprecationWarning:
    Attribute s is deprecated and will be removed in Python 3.14; use value instead
    string = fmt_call.func.value.s
```
```python
  lib/python3.12/ast.py:407: DeprecationWarning:
    visit_Str is deprecated; add visit_Constant
    return visitor(node)
```

**Comment by ikamensh** [comment]
Thanks for reporting!
hm, it seems ast.Str and ast.Num and perhaps other nodes will be all merged into ast.Constant. Naively replacing ast.Str with ast.Constant leads to errors in flynt.

**Comment by akaihola** [comment]
This currently blocks akaihola/darker#774 from enabling CI tests with Python 3.14-dev.

**Comment by ikamensh** [comment]
Ok python 3.14 is on the horizon now, so this gains in importance. PR would be welcome.

**Comment by ikamensh** [comment]
`isinstance(x, ast.Str)` should roughly translate as `isinstance(x, ast.Constant) and isinstance(x.value, str)`, however when I tried this replacement some cases didn't work at first. I have to say I've spent little time on it though. Perhaps if you go one by one in finding usage of deprecated classes you will find solution for each case. I'd say unless you have reason to suspect otherwise we can assume test suite coverage to be enough, so if you change classes and tests still pass it should be good to go. 

It will also be good if it comes as series of PRs, you can leave complicated cases (if any) for later.

**Comment by akaihola** [comment]
> migration from `astor` to `ast.unparse()` needs to be done in `ast_to_string()`

...and support for Python 3.8 (end-of-life currentl) needs to be dropped.

**Comment by akaihola** [comment]
> expect double quotes instead of single quotes

@ikamensh is it ok if the behavior of Flynt changes with regard to single and double quotes when moving from `astor` to `ast.unparse()`?

**Comment by ikamensh** [comment]
Can you give some examples? The prob is if it gives big change list on projects already having flynt as linter. You can see projects in "used by" and run new version on their code. If it changes every other file, it can be quite annoying.

**Comment by ikamensh** [comment]
I can imagine behavior change via minor version bump, if new behavior is sensible

**Comment by akaihola** [comment]
> gives big change list on projects already having flynt as linter

I don't believe it would do that, since already converted f-strings are not considered by Flynt, no matter what kind of quotes they use.

But it *is* annoying that [`ast.unparse()`] outputs single quoted strings by default while modern code reformatters default to double quotes.

[`ast.unparse()`] behavior is impossible to change since it:
- doesn't provide any options to tweak its behavior
- uses the [`ast._Unparser`] class which is not part of the public API

To preserve the current double quote behavior, I can think of three options:
1. help [`astor`] support Python 3.14
2. vendor and patch [`ast._Unparser`] and [`ast.unparse()`]
3. post-process unparsed code using Black

[`astor`]: https://pypi.org/project/astor
[`ast._Unparser`]: https://github.com/python/cpython/blob/5fa7e1b7fd57e8c6297e9eb79d79cede42e5ce0f/Lib/ast.py#L666
[`ast.unparse()`]: https://docs.python.org/3/library/ast.html#ast.unparse

**Comment by ikamensh** [comment]
I think option 3 is promising.

**Comment by akaihola** [comment]
@ikamensh, also option 4: contribute to [astunparse] ([on GitHub]) by updating the current copy of [`Tools/parser/unparse.py` from Python 3.8.0] to code for [`unparse()` in Python 3.14a4] and adding an optional argument for preferred quoting of string literals – provided that @simonpercivall is willing to accept a patch and make a new release.

And option 5: do the above in a new PyPI package.

[astunparse]: https://pypi.org/project/astunparse/
[on GitHub]: /simonpercivall/astunparse
[`Tools/parser/unparse.py` from Python 3.8.0]: /python/cpython/blob/v3.8.0/Tools/parser/unparse.py
[`unparse()` in Python 3.14a4]: /python/cpython/blob/f26daa9470925120f6336ca508f7ea193b00a3f4/Lib/ast.py#L666-L1740

**Comment by ikamensh** [comment]
A new package needs to be maintained, I would first explore other approaches

**Comment by ikamensh** [comment]
I am already using black in some places

**Comment by akaihola** [comment]
To summarize discussion from #205 also here, the conclusion for preserving the current double quote behavior was to

> 1. help [`astor`](https://pypi.org/project/astor) support Python 3.14

(see berkerpeksag/astor#217)

**Comment by ikamensh** [comment]
in #219 I've removed the dependency on astor. Also usage of deprecated ast classes in the code was migrated.
